[JW8-1288] Reset autostart cancelable on pause.

vseventer opened the pull request:
### This PR will...
Reset the autostart cancelable on pause.

### Why is this Pull Request needed?
~~On mobile, if the element is not viewable (i.e. added to the DOM only after ready), we currently autopause the video, which cancels the autostart check. Once the element becomes viewable (i.e. is added to the DOM), however, we now want to perform the canAutoplay test - but because the autostart was cancelled earlier, this will fail. By resetting the autostart cancelable, the test always uses the latest cancelable, which will work as expected.~~

Resets cancelable before every test instead instead of on pause. In addition, moved the `autostartMuted` change listener to after we set it for the first time. Finally, once a user clicks-to-play, we never autostart/pause on viewable change (similar to click-to-pause behavior).

### Are there any points in the code the reviewer needs to double check?
~~Alternatively, I looked into only calling _pause is the element is not currently idle - however this is a bit more complicated as there are other non-playing states (e.g. buffering / error). This seems to be the cleaner solution.~~

#### Addresses Issue(s):
JW8-1288

johnBartos commented:
<!--
  0 failure: 
  1 warning:  🛠 There are modi...
  
  
  DangerID: danger-id-default;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  🛠 There are modified src files, but no test changes. Add tests if you're able to.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

robwalch commented on src/js/controller/controller.js:
```diff
@@ -486,6 +486,7 @@ Object.assign(Controller.prototype, {
         function _pause(meta) {
             _actionOnAttach = null;
             checkAutoStartCancelable.cancel();
+            checkAutoStartCancelable = cancelable(_checkAutoStart);
```
@vseventer I notice we're also doing this on `_load()`, but not not when cancelling on `_play()` and `_stop()`.

The reason we reset the cancelable on load is because loading a new playlist results in an autostart check. play, stop and even pause should cancel autostart behavior.

If you pause while the player is attempting to start, it shouldn't start, or at least it should start and then pause. I'm not sure that's what will happen with this change.

egreaves replied:
@robwalch would it suffice to check the pause reason here to see if it's from a viewability change and not a user interaction or api call to `pause()`? This may be simpler than changing `_checkPlayOnViewable`.
```
if (pauseReason === 'autostart') {
    checkAutoStartCancelable = cancelable(_checkAutoStart);
}
```

robwalch replied:
I don't think so. We should look at the timing of event listeners. There is not reason to pause on "viewable" until playback has started, so splitting up the "viewable" event handler to one that autostarts, and another that pauses once playback has started is the right move.

vseventer commented:
@robwalch Fair enough, we can move it to `_checkPlayOnViewable`, but keep in mind we still need to reset the cancelable to allow for playback to pause -> resume -> pause -> resume etc. when scrolled in- and out of view without interaction.

vseventer commented:
@robwalch @egreaves I'm revisiting this ticket altogether - there are some issues with viewable mainly on mobile, mostly caused by the cancelable and autostartMuted change listener which fires prematurely since the autoplay test is now async.

robwalch commented:
@vseventer (+ @egreaves ) Have you already started on this? I think I have a good solution.

The essence of the problem is that pause is called while we're trying to autostart (or test capabilities). 
```
function _playerReadyNotify() {
...
    _model.change('viewable', _checkPlayOnViewable);
...
        function _checkPlayOnViewable(model, viewable) {
            if (model.get('playOnViewable')) {
                if (viewable) {
                    _autoStart();
                } else if (OS.mobile) {
                    _this.pause({ reason: 'autostart' });
                }
            }
        }
```

We always want this pause behavior for mobile autostart so we just need to split this logic out:
function _playerReadyNotify() {
...
    _model.change('viewable', _checkPlayOnViewable);
...
        function _checkPlayOnViewable(model, viewable) {
            if (model.get('playOnViewable')) {
                if (viewable) {
                    _autoStart();
            }
        }
```

Then in `_autoStart` after the test if we're `OS.mobile` implement play/pause toggling on viewable, until or unless the state of these has changed:`_model.once('change:autostartFailed change:autostartMuted change:mute', function(model) {`.

vseventer commented:
Updated PR to reset cancelable before every test instead of on pause. In addition, moved the `autostartMuted` change listener to after we set it for the first time. Finally, once a user clicks-to-play, we never autostart/pause on viewable change (similar to click-to-pause behavior).

robwalch commented on src/js/controller/controller.js:
```diff
@@ -376,6 +376,10 @@ Object.assign(Controller.prototype, {
+            // Stop autoplay behavior if the video is started by the user or an api call
+            if (playReason === 'interaction' || playReason === 'external') {
+                _model.set('playOnViewable', false);
+            }
```
This is fine for now, but we need to make a tech-debt note that model properties shouldn't span view-controls code and player-model-controller. Everything "autostart" and "playOnViewable" is split up between controls and controller instead of being encapsulated in one place, like a module that listens to model changes ("change:playReason" where reason === "interaction").

robwalch commented on src/js/controller/controller.js:
```diff
@@ -435,6 +442,10 @@ Object.assign(Controller.prototype, {
                     _model.set('autostartMuted', true);
                     updateProgramSoundSettings();
+
+                    _model.once('change:autostartMuted', function(model) {
```
Better to do this right after the `set`
